File: MATTHEW/leetcode/AIcoding/3.DeepSeek-v3/rust_solutions/2558-take-gifts-from-the-richest-pile.rs
use std::collections::BinaryHeap;

impl Solution {
    pub fn pick_gifts(gifts: Vec<i32>, k: i32) -> i64 {
        let mut heap = BinaryHeap::from(gifts);
        for _ in 0..k {
            if let Some(max) = heap.pop() {
                let remaining = (max as f64).sqrt().floor() as i32;
                heap.push(remaining);
            }
        }
        heap.into_iter().map(|x| x as i64).sum()
    }
}

File: leetcode/AIcoding/3.DeepSeek-v3/erlang_solutions/2558-take-gifts-from-the-richest-pile.erl
-module(solution).
-export([pick_gifts/2]).

pick_gifts(Gifts, K) ->
    pick_gifts(Gifts, K, lists:sort(fun(A, B) -> A >= B end, Gifts)).

pick_gifts(Gifts, 0, _) ->
    lists:sum(Gifts);
pick_gifts(Gifts, K, [Max | Rest]) ->
    NewMax = math:sqrt(Max),
    NewGifts = lists:sort(fun(A, B) -> A >= B end, [trunc(NewMax) | Rest]),
    pick_gifts(NewGifts, K - 1, NewGifts).